src/lib/token.test.ts

```diff
@@ -0,0 +1,349 @@
+  let Local = Mina.LocalBlockchain();
+  Mina.setActiveInstance(Local);
+  feePayer = Local.testAccounts[0].privateKey;
+
+  zkappKey = PrivateKey.random();
+  zkappAddress = zkappKey.toPublicKey();
+  zkapp = new TokenContract(zkappAddress);
+
+  tokenAccount1Key = Local.testAccounts[1].privateKey;
+  tokenAccount1 = tokenAccount1Key.toPublicKey();
+
+  tokenAccount2Key = Local.testAccounts[2].privateKey;
+  tokenAccount2 = tokenAccount2Key.toPublicKey();
+
+  (
+    await Mina.transaction(feePayer, () => {
+      Party.fundNewAccount(feePayer);
+      zkapp.deploy({ zkappKey });
+      zkapp.init();
+    })
+  ).send();
+}
+
+describe('Token', () => {
+  beforeAll(async () => await isReady);
+
+  afterAll(() => setTimeout(shutdown, 0));
+
+  describe('Create existing token', () => {
+    beforeEach(async () => {
+      await setupLocal();
+    });
+
+    it('should have a valid custom token id', async () => {
+      const tokenId = zkapp.token().id;
+      const expectedTokenId = new Token({ tokenOwner: zkappAddress }).id;
+      expect(tokenId).toBeDefined();
+      expect(tokenId).toEqual(expectedTokenId);
+    });
+
+    it('should have a valid token symbol', async () => {
+      const symbol = Mina.getAccount(zkappAddress).tokenSymbol;
+      expect(tokenSymbol).toBeDefined();
+      expect(symbol).toEqual(tokenSymbol);
+    });
+
+    it('should create a valid token with a different parentTokenId', async () => {
+      const newTokenId = Ledger.customTokenId(tokenAccount1, zkapp.token().id);
+      expect(newTokenId).toBeDefined();
+    });
+
+    it('should error if passing in an invalid tokenSymbol', async () => {
+      await Mina.transaction(feePayer, () => {
+        zkapp.setInvalidTokenSymbol();
+        zkapp.sign(zkappKey);
+      }).catch((e) => {
+        expect(e).toBeDefined();
+      });
+    });
+  });
+
+  describe('Mint token', () => {
+    beforeEach(async () => {
+      await setupLocal();
+    });
+
+    it('should change the balance of a token account after token owner mints', async () => {
+      (
+        await Mina.transaction(feePayer, () => {
+          Party.fundNewAccount(feePayer);
+          zkapp.mint(tokenAccount1, UInt64.from(100_000));
+          zkapp.sign(zkappKey);
+        })
+      ).send();
+
+      expect(
+        Mina.getBalance(tokenAccount1, zkapp.token().id).value.toBigInt()
+      ).toEqual(BigInt(100_000));
+    });
+
+    it('should error if token owner mints more tokens than allowed', async () => {
+      await Mina.transaction(feePayer, () => {
+        Party.fundNewAccount(feePayer);
+        zkapp.mint(tokenAccount1, UInt64.from(100_000_000_000));
+        zkapp.sign(zkappKey);
+      }).catch((e) => {
+        expect(e).toBeDefined();
+      });
+    });
+
+    it('should error if token owner mints negative amount tokens', async () => {
+      await Mina.transaction(feePayer, () => {
+        Party.fundNewAccount(feePayer);
+        zkapp.mint(tokenAccount1, UInt64.from(-100_000));
```

Comment:
this throws an error because `UInt64.from(<negative number>)` isn't possible, so it's not exercising the test

Reply:
The accepted types for `mint`, `burn` and `send` are all UInts so I guess this test doesn't do a great job at its goal. I removed these tests since it's really a check on how UInt behaves